### PR TITLE
Improve data pipeline for punctuation capitalization model and make other useful changes

### LINKED ISSUE #3125: Punctuation data set uses too much memory
**Describe the bug**

Punctuation datasets cannot be constructed with more than a few million examples (depending on the max length). Even these small datasets can consume a huge amount of memory while preprocessing (100's of GB) and while training (10's of GB per process). 

This issue was mentioned in passing in https://github.com/NVIDIA/NeMo/issues/2569 though it was not the main issue discussed there and thus not fixed.

Assuming it is an acceptable idea to load the entire dataset into memory in the first place, these are the points where there seems to be an excessive use of memory.

The function with issues (`get_features`):
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py#L34 

The main issues:
1. Why save `all_segment_ids`? It appears there is nowhere this is set to anything but a constant 0. `__getitem__` can deal with it.
2. Why pad to the max length? `collate_fn` can deal with that, and could do so much more efficiently (by only padding to the max length of the batch).
3. Loss mask and input mask can be generated by `__getitem__` or `collate_fn`
4. Subtokens mask could be generated later as well, albeit not trivially.

**Expected behavior**

Should be able to use a dataset with 10's of millions of lines.

### PR DESCRIPTION
**Pull request by PeganovAnton**
1. `PunctuationCapitalizationDataset`
  - Sort input sequences before packing them into batches to reduce padding
  -  Batching is done in `PunctuationCapitalizationDataset`, not in `DataLoader`
  -  Parallelize features creation
  -  Exclude `'input_mask'`, `'loss_mask'`, `'segment_ids'` from dataset cache. Now these arrays are created in dataset constructor
  - `PunctuationCapitalizationDataset` now yields dictionaries, not tuples
2. Tarred dataset class and instruments for tarred dataset creation.
3. Classes for punctuation and capitalization configs
4. Now parameters `max_seq_length`, `use_cache`, `num_workers`, `pin_memory`, `drop_last` can be passed in `train_ds`, `validation_ds`, `test_ds` and in these sections they have higher priority.
5. Add separate metrics for every evaluation dataset.

Closes #3125

### CODE REVIEW

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py [hunk 1]
```diff
@@ -12,209 +12,682 @@
...
+        self.punct_label_ids = punct_label_ids
+        self.capit_label_ids = capit_label_ids
+        self.pad_label = pad_label
+        self.with_label = with_label
+        self.verbose = verbose
+        self.progress_queue = progress_queue
+
+    def maybe_clip(self, values: List[int], prepend_value: int) -> List[int]:
+        if len(values) > self.max_seq_length:
+            return [prepend_value] + values[-self.max_seq_length + 1 :]
```
Comment (ekmb) [inline nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py:303]: why do you choose to use the end?

**Review by ekmb — COMMENTED** [review]
initial round

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml [hunk 2]
```diff
@@ -58,31 +58,37 @@ model:
...
 
   train_ds:
+    use_tarred_dataset: false
     text_file: text_train.txt
     labels_file: labels_train.txt
+    metadata_file: null
     shuffle: true
     num_samples: -1
-    batch_size: 64
+    tokens_in_batch: 15000
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:67]: Grouped tarred dataset parameters

[the remark this answers is not in the document]
Comment (PeganovAnton) [reply examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:67]: Using `tokens_in_batch` may give a significant boost. After I started to use this parameter, number of steps in epoch became 5-6 times smaller. There also was a significant improvement of metrics.

File: examples/nlp/token_classification/data/create_punctuation_capitalization_tarred_dataset.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        "--tokenizer_model", "-m", type=Path, help="Path to tokenizer model required for 'sentencepiece' tokenizer."
+    )
+    parser.add_argument(
+        "--vocab_file",
+        "-v",
+        type=Path,
+        help="Path to vocabulary file which is used in 'word', 'char', and HuggingFace tokenizers.",
+    )
+    parser.add_argument(
+        "--merges_file", "-M", type=Path, help="Path to merges file which maybe used in HuggingFace tokenizers."
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply examples/nlp/token_classification/data/create_punctuation_capitalization_tarred_dataset.py:130]: For possible usage of all of other tokenizers. It looks like some HuggingFace tokenizers require it.

File: examples/nlp/token_classification/data/create_punctuation_capitalization_tarred_dataset.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    )
+    parser.add_argument(
+        "--max_seq_length",
+        "-s",
+        type=int,
+        default=512,
+        help="Maximum number of subtokens in an input sequence. A source sequence which contain too many subtokens are "
+        "clipped to `--max_seq_length - 2` subtokens and then [CLS] token is prepended to the clipped sequence and "
+        "[SEP] token is appended to the clipped sequence. The clipping is performed via removal of subtokens in the "
+        "beginning of a source sequence.",
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply examples/nlp/token_classification/data/create_punctuation_capitalization_tarred_dataset.py:84]: I kept removal from the beginning as it was done before. Frankly, I do not know why is it done this way.

File: examples/nlp/token_classification/punctuation_capitalization_evaluate.py [hunk 5]
```diff
@@ -56,7 +60,8 @@ def main(cfg: DictConfig) -> None:
         'During evaluation/testing, it is currently advisable to construct a new Trainer with single GPU and \
             no DDP to obtain accurate results'
     )
-
+    default_cfg = PunctuationCapitalizationConfig()
+    cfg = update_model_config(default_cfg, cfg)
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply examples/nlp/token_classification/punctuation_capitalization_evaluate.py:64]: Checking errors in parameters, that user provides

File: examples/nlp/token_classification/punctuation_capitalization_train.py [hunk 6]
```diff
@@ -61,6 +66,9 @@
 
 @hydra_runner(config_path="conf", config_name="punctuation_capitalization_config")
 def main(cfg: DictConfig) -> None:
+    torch.manual_seed(42)
+    default_cfg = PunctuationCapitalizationConfig()
+    cfg = update_model_config(default_cfg, cfg)
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply examples/nlp/token_classification/punctuation_capitalization_train.py:71]: Checking errors in parameters, that user provides

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py [hunk 7]
```diff
@@ -141,110 +141,98 @@ def training_step(self, batch, batch_idx):
...
-            'punct_fp': self.punct_class_report.fp,
-            'capit_tp': self.capit_class_report.tp,
-            'capit_fn': self.capit_class_report.fn,
-            'capit_fp': self.capit_class_report.fp,
-        }
+        return self.eval_step(batch, 'test', dataloader_idx)
+
+    def training_epoch_end(self, outputs: EPOCH_OUTPUT) -> None:
+        shuffle_train_dataset = self._cfg.dataset.get('shuffle_train_dataset')
+        if shuffle_train_dataset is None:  # Encountered legacy config
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py]: If default is not set, then it is `None`.

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py [hunk 8]
```diff
@@ -141,110 +141,98 @@ def training_step(self, batch, batch_idx):
...
+            shuffle_train_dataset = not self.cfg.train_ds.get('use_tarred_dataset', False)
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py]: I keep it for shuffling batches and for repacking batches there is a separate parameter. Do you

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_tarred_dataset.py [hunk 9]
```diff
@@ -0,0 +1,968 @@
...
+    capit_label_ids = output_dir / DEFAULT_CAPIT_LABEL_IDS_NAME
+    if punct_label_ids.exists():
+        logging.warning(f"Found unexpected punctuation label file {punct_label_ids}. It is going to be removed.")
+        punct_label_ids.unlink()
+    if capit_label_ids.exists():
+        logging.warning(f"Found unexpected capitalization label file {capit_label_ids}. It is going to be removed.")
+        capit_label_ids.unlink()
+
+
+def collect_unique_labels_from_fragment(
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply nemo/collections/nlp/data/token_classification/punctuation_capitalization_tarred_dataset.py:280]: The `collect_unique_labels_from_fragment` function is called from `create_label_dictionaries` for all fragments and then labels are united.

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_tarred_dataset.py [hunk 10]
```diff
@@ -0,0 +1,968 @@
...
+            )
+            for fragment_idx, start_pos in enumerate(text_start_bytes)
+        )
+    unique_punct, unique_capit = zip(*result)
+    unique_punct = set().union(*unique_punct)
+    unique_capit = set().union(*unique_capit)
+    return create_label_ids(unique_punct, pad_label), create_label_ids(unique_capit, pad_label)
+
+
+def check_label_ids(pad_label: str, punct_label_ids: Dict[str, int], capit_label_ids: Dict[str, int]):
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply nemo/collections/nlp/data/token_classification/punctuation_capitalization_tarred_dataset.py:359]: The option of passing ready label dictionary to the script is present. See parameters `--punct_labels`, `--capit_labels`, `--pad_label` and `--punct_label_vocab_file`, `--capit_label_vocab_file`.

In the future I plan to train models on more than 3 labels and in such a case creating label ids on the fly may be useful.

File: nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py (hunk 1 above)
Comment (PeganovAnton) [reply nemo/collections/nlp/data/token_classification/punctuation_capitalization_dataset.py:320]: I do not know.

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml [hunk 11]
```diff
@@ -39,50 +41,84 @@ exp_manager:
...
     ignore_start_end: false
-    use_cache: true
-    # shared among dataloaders
-    num_workers:  2
-    pin_memory: false
-    drop_last: false
+    punct_label_ids: null
+    punct_label_vocab_file: null
+    capit_label_ids: null
+    capit_label_vocab_file: null
```
Comment (ekmb) [inline examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:51]: please revert changes removing class_labels.

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml [hunk 12]
```diff
@@ -39,50 +41,84 @@ exp_manager:
...
+
+    # Path to tarred dataset metadata file. Required if tarred dataset is used. Metadata file is a JSON file which
+    # contains total number of batches in the dataset, a list of paths to tar files and paths to label vocabularies.
+    # Metadata file is create by script
+    # `examples/nlp/token_classification/create_punctuation_capitalization_tarred_dataset.py`
+    tar_metadata_file: null
+    # Controls batch shuffling in tarred dataset. `tar_shuffle_n` is a size of shuffled batch buffer. Mind that this
+    # shuffling only permutes batches and doesn't exchange samples between batches. Proper shuffling is turned on in
+    # usual dataset.
+    tar_shuffle_n: 1
```
Comment (ekmb) [inline examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:81]: How about moving tar_shuffle_n, tart_metadata_file, and n_jobs right after use_tarred - with the note that these are only for tarred dataset, tokens_n_batch - for both tar and regular dataset, the rest args for regular format.

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekmb) [reply examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:67]: Could you add a comment here on what this arg means?

File: nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py [hunk 13]
```diff
@@ -273,71 +295,161 @@ def setup_training_data(self, train_data_config: Optional[DictConfig] = None):
...
-        # use data_dir specified in the ds_item to run evaluation on multiple datasets
-        if 'ds_item' in cfg and cfg.ds_item is not None:
-            data_dir = cfg.ds_item
+        ds_item = cfg.get('ds_item')
+        data_dir = self._cfg.dataset.get('data_dir')
+        # Following parameters can be missing in config if the model is restored from old checkpoint
+        use_tarred_dataset = cfg.get('use_tarred_dataset', False)
+        num_workers = cfg.get('num_workers')
+        pin_memory = cfg.get('pin_memory')
+        drop_last = cfg.get('drop_last')
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply nemo/collections/nlp/models/token_classification/punctuation_capitalization_model.py]: There was a major refactoring. I think this one is outdated.

File: Jenkinsfile [hunk 14]
```diff
@@ -1087,13 +1087,70 @@ pipeline {
...
-            pretrained_model=/home/TestData/nlp/token_classification_punctuation/output/checkpoints/Punctuation_and_Capitalization.nemo && \
+            pretrained_model=/home/TestData/nlp/token_classification_punctuation/output/checkpoints/Punctuation_and_Capitalization.nemo \
+            model.dataset.data_dir=/home/TestData/nlp/token_classification_punctuation/ && \
             rm -rf /home/TestData/nlp/token_classification_punctuation/output/*'
           }
         }
       }
+      post {
+        always {
+          sh 'pwd && ls nemo_* && rm -rf nemo_experiments && ls nemo_*'
```
[the remark this answers is not in the document]
Comment (PeganovAnton) [reply Jenkinsfile:1150]: These tests do not conflict when they are run. Very weird problem emerged when several stages tried to remove experiment directory simultaneously.

File: Jenkinsfile [hunk 15]
```diff
@@ -1068,35 +1076,138 @@ pipeline {
...
+              model.test_ds.ds_item=tmp_data \
+              model.language_model.pretrained_model_name=distilbert-base-uncased \
+              +model.train_ds.use_cache=false \
+              +model.validation_ds.use_cache=false \
+              +model.test_ds.use_cache=false \
+              trainer.gpus=[0,1] \
+              trainer.accelerator=ddp \
+              trainer.max_epochs=1 \
+              +exp_manager.explicit_log_dir=/home/TestData/nlp/token_classification_punctuation/output \
+              +do_testing=true && \
```
Comment (ekmb) [inline Jenkinsfile:1112]: with this new `do_testing` flag, `punctuation_capitalization_evaluate.py` can be deleted, maybe move its inference part to the `punctuate_capitalize_infer.py`

File: docs/source/nlp/punctuation_and_capitalization.rst [hunk 16]
```diff
@@ -158,55 +162,702 @@ for the training and evaluation in the :code:`conf/punctuation_capitalization_co
...
+   :header-rows: 1
+
+   * - **Parameter**
+     - **Data type**
+     - **Default value**
+     - **Description**
+   * - **class_labels**
+     - :ref:`class labels config<class-labels-config-label>`
+     - :ref:`class labels config<class-labels-config-label>`
+     - A mandatory parameter containing a dictionary with names of label id files used in ``.nemo`` checkpoints. These
```
Comment (ekmb) [inline docs/source/nlp/punctuation_and_capitalization.rst:287]: I'd reword this part. Users don't need to pass anything to this args, but they're required to create a proper .nemo file with all artifacts. Or do you mean smth different?

File: docs/source/nlp/punctuation_and_capitalization.rst [hunk 17]
```diff
@@ -158,55 +162,702 @@ for the training and evaluation in the :code:`conf/punctuation_capitalization_co
...
+   :widths: 5 5 5 35
+   :header-rows: 1
+
+   * - **Parameter**
+     - **Data type**
+     - **Default value**
+     - **Description**
+   * - **punct_labels_file**
+     - string
+     - ???
```
Comment (ekmb) [inline docs/source/nlp/punctuation_and_capitalization.rst:354]: the config has a default value for this param

File: docs/source/nlp/punctuation_and_capitalization.rst [hunk 18]
```diff
@@ -158,55 +162,702 @@ for the training and evaluation in the :code:`conf/punctuation_capitalization_co
...
+     - A name of a punctuation labels file. It is a mandatory parameter that cannot be omitted in the config. This name
+       is used as a name of label ids file in ``.nemo`` checkpoint. It also can be used for passing label vocabulary to
+       the model. If ``punct_labels_file`` is used as a vocabulary file, then you should provide parameter
+       ``label_vocab_dir`` in :ref:`common dataset parameters<common-dataset-parameters-config-label>`
+       (``model.common_dataset_parameters.label_vocab_dir`` in :ref:`run config<run-config-label>`). Each line of
+       ``punct_labels_file`` file contains 1 label. The first line has to contain neutral label which has to be
+       equal to a ``pad_label`` parameter in :ref:`common dataset parameters<common-dataset-parameters-config-label>`.
+   * - **capit_labels_file**
+     - string
+     - ???
```
Comment (ekmb) [inline docs/source/nlp/punctuation_and_capitalization.rst:364]: the config has a default value for this param

File: docs/source/nlp/punctuation_and_capitalization.rst [hunk 19]
```diff
@@ -158,55 +162,702 @@ for the training and evaluation in the :code:`conf/punctuation_capitalization_co
...
+   * - **punct_labels_file**
+     - string
+     - ???
+     - A name of a punctuation labels file. It is a mandatory parameter that cannot be omitted in the config. This name
+       is used as a name of label ids file in ``.nemo`` checkpoint. It also can be used for passing label vocabulary to
+       the model. If ``punct_labels_file`` is used as a vocabulary file, then you should provide parameter
+       ``label_vocab_dir`` in :ref:`common dataset parameters<common-dataset-parameters-config-label>`
+       (``model.common_dataset_parameters.label_vocab_dir`` in :ref:`run config<run-config-label>`). Each line of
+       ``punct_labels_file`` file contains 1 label. The first line has to contain neutral label which has to be
+       equal to a ``pad_label`` parameter in :ref:`common dataset parameters<common-dataset-parameters-config-label>`.
```
Comment (ekmb) [inline docs/source/nlp/punctuation_and_capitalization.rst:361]: the values are sorted, line number==label id, starting from 0

File: docs/source/nlp/punctuation_and_capitalization.rst [hunk 20]
```diff
@@ -158,55 +162,702 @@ for the training and evaluation in the :code:`conf/punctuation_capitalization_co
...
+   :header-rows: 1
+
+   * - **Parameter**
+     - **Data type**
+     - **Default value**
+     - **Description**
+   * - **pad_label**
+     - string
+     - ???
+     - A mandatory parameter which should contain label used for punctuation and capitalization label padding. It
```
Comment (ekmb) [inline docs/source/nlp/punctuation_and_capitalization.rst:401]: there is a default value as well - I would avoid the word `mandatory` for config args with default values that users wouldn't want to change.

File: docs/source/nlp/punctuation_and_capitalization.rst [hunk 21]
```diff
@@ -158,55 +162,702 @@ for the training and evaluation in the :code:`conf/punctuation_capitalization_co
...
+   * - :ref:`Run config<run-config-label>`
+     - ``model.train_ds``, ``model.validation_ds``, ``model.test_ds``
+   * - :ref:`Model config<model-config-label>`
+     - ``train_ds``, ``validation_ds``, ``test_ds``
+
+For convenience, items of data config are described in 4 tables:
+:ref:`common parameters for both regular and tarred datasets<common-data-parameters-label>`,
+:ref:`parameters which are applicable only to regular dataset<regular-dataset-parameters-label>`,
+:ref:`parameters which are applicable only to tarred dataset<tarred-dataset-parameters-label>`,
+:ref:`parameters for PyTorch data loader<pytorch-dataloader-parameters-label>`.
```
Comment (ekmb) [inline docs/source/nlp/punctuation_and_capitalization.rst:453]: Thanks for adding all these details!

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml [hunk 22]
```diff
@@ -39,50 +41,88 @@ exp_manager:
...
 model:
-  punct_label_ids: null # will be filled during training
-  capit_label_ids: null # will be filled during training
   class_labels:
     punct_labels_file: punct_label_ids.csv
     capit_labels_file: capit_label_ids.csv
-  dataset:
-    data_dir: ??? # /path/to/training/data
-    # parameters for dataset preprocessing
-    max_seq_length: 128
```
Comment (ekmb) [inline examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:50]: could you add back max_seq_length

File: docs/source/nlp/punctuation_and_capitalization.rst [hunk 23]
```diff
@@ -158,55 +162,702 @@ for the training and evaluation in the :code:`conf/punctuation_capitalization_co
...
+     - A mandatory parameter if ``use_tarred_dataset=false``. ``labels_file`` is a name of a file with punctuation and
+       capitalization labels in :ref:`NeMo format <nemo-data-format-label>`. It has is located in ``ds_item`` directory.
+   * - **tokens_in_batch**
+     - int
+     - ``null``
+     - A mandatory parameter if ``use_tarred_dataset=false``. ``tokens_in_batch`` contains a number of tokens in a batch
+       including paddings and special tokens ([CLS], [SEP], [UNK]). This config does not have ``batch_size`` parameter.
+   * - **max_seq_length**
+     - int
+     - ``512``
```
Comment (ekmb) [inline docs/source/nlp/punctuation_and_capitalization.rst:516]: 128? please check the default values, they don't match the config

File: examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml (hunk 22 above)
Comment (PeganovAnton) [reply examples/nlp/token_classification/conf/punctuation_capitalization_config.yaml:50]: It is now in `train_ds`, `validation_ds`, `test_ds`. I added it back these sections

**Review by ekmb — APPROVED** [review]

**Review by ekmb — APPROVED** [review]
